engines/ic3base.cpp

```diff
@@ -399,98 +390,112 @@ bool IC3Base::rel_ind_check(size_t i,
 
 bool IC3Base::block_all()
 {
-  while (has_proof_goals()) {
-    if (options_.ic3_reset_interval_
-        && num_check_sat_since_reset_ >= options_.ic3_reset_interval_) {
-      reset_solver();
-    }
-
-    const ProofGoal * pg = get_top_proof_goal();
-    if (is_blocked(pg)) {
-      logger.log(3,
-                 "Skipping already blocked proof goal <{}, {}>",
-                 pg->target.term->to_string(),
-                 pg->idx);
-      remove_top_proof_goal();
-      continue;
-    };
-
-    // block can fail, which just means a
-    // new proof goal will be added
-    if (block(pg)) {
-      // if successfully blocked, then remove that proof goal
-      // expecting the top proof goal to still be pg
-      assert(pg == get_top_proof_goal());
-      remove_top_proof_goal();
-    } else if (!pg->idx) {
-      // if a proof goal cannot be blocked at zero
-      // then there's a counterexample
-      // NOTE: creating a new allocation
-      //       because the pg memory is already managed
-      //       by proof_goals_
-      cex_pg_ = new ProofGoal(pg->target, pg->idx, pg->next);
-      return false;
-    }
-  }
-  assert(!has_proof_goals());
-  return true;
-}
-
-bool IC3Base::block(const ProofGoal * pg)
-{
-  const IC3Formula & c = pg->target;
-  size_t i = pg->idx;
+  assert(!solver_context_);
+  ProofGoalQueue proof_goals;
+  IC3Formula bad_goal;
+  while (intersects_bad(bad_goal)) {
+    assert(bad_goal.term);  // expecting non-null
+    assert(proof_goals.empty());  // bad should be the first goal each iteration
+    proof_goals.new_proof_goal(bad_goal, frontier_idx(), nullptr);
+
+    while (!proof_goals.empty()) {
+      const ProofGoal * pg = proof_goals.top();
+
+      if (!pg->idx) {
+        // went all the way back to initial
+        // TODO refactor refinement to not use cex_pg_
+        // need to create a new proof goal that's not managed by the queue
+        cex_pg_ = new ProofGoal(pg->target, pg->idx, pg->next);
+        RefineResult s = refine();
+        if (s == REFINE_SUCCESS) {
+          // on successful refinement, clear the queue of proof goals
+          // which might not have been precise
+          // TODO might have to change this if there's an algorithm
+          // that refines but can keep proof goals around
+          proof_goals.clear();
+
+          // and reset cex_pg_
+          if (cex_pg_) {
+            delete cex_pg_;
+            cex_pg_ = nullptr;
+          }
+          continue;
```

Comment:
this is not needed

Reply:
The `continue`? Actually it is because otherwise the rest of the loop body will run on an old proof goal.

Reply:
But that's one of the things I can refactor later to make more clear

Reply:
right. I think I didn't fully see the indentation